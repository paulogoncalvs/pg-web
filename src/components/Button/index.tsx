import classNames from 'classnames';
import { h, FunctionalComponent, ComponentChildren } from 'preact';

interface ButtonComponentProps {
    mainClasses?: string;
    classes?: string;
    otherProps?: unknown;
    children?: ComponentChildren;
}

export const Button: FunctionalComponent<ButtonComponentProps> = ({
    mainClasses = 'btn',
    classes,
    children,
    ...otherProps
}) => (
    <button class={classNames(mainClasses, classes)} {...otherProps}>
        {children}
    </button>
);
