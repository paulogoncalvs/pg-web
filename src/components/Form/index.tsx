import { h, FunctionalComponent, ComponentChildren, createContext } from 'preact';
import { useState } from 'preact/hooks';

export interface FormComponentData {
    [key: string]: string;
}
interface FormComponentProps {
    action: string;
    initialValues: FormComponentData;
    onSubmit(formData: FormComponentData): void;
    classes: string;
    children: ComponentChildren;
}

export const FormContext = createContext({
    formData: {},
    handleFormChange: (event: Event): void => console.warn(event), // eslint-disable-line no-console
});

export const Form: FunctionalComponent<FormComponentProps> = ({
    action,
    initialValues,
    onSubmit,
    classes,
    children,
}) => {
    const [formData, setFormData] = useState(initialValues);

    const handleFormChange = (event: Event): void => {
        const { name, value } = event.target as HTMLInputElement;

        setFormData({ ...formData, [name]: value });
    };

    const triggerSubmit = (event: Event): void => {
        if (!onSubmit) {
            return;
        }

        event.preventDefault();
        onSubmit(formData);
    };

    return (
        <form action={action} onSubmit={triggerSubmit} class={classes}>
            <FormContext.Provider value={{ formData, handleFormChange }}>{children}</FormContext.Provider>
        </form>
    );
};
