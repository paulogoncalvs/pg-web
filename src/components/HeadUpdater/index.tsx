import routesConfig from '@/config/routes/index.js';
import { useRouter } from '@/modules/router';
import { isBrowser } from '@/utils/browser';

export const HeadUpdater = (): null => {
    useRouter();

    if (!isBrowser()) {
        return null;
    }

    const data = routesConfig[window.location.pathname.replace(/\/?$/, '/')] || {};

    if (data.templateParameters) {
        const { title } = data.templateParameters.head;

        document.title = title;
    }

    return null;
};
