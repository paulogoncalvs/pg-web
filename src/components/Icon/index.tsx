import { h, FunctionalComponent } from 'preact';
import { useContext } from 'preact/hooks';
import { AppContext } from '@/app/AppContext';

interface IconComponentProps {
    src: IconSrc;
    width?: string;
    height?: string;
    classes?: string;
    viewBox?: string;
    otherProps?: unknown;
}

export const Icon: FunctionalComponent<IconComponentProps> = ({
    src: { viewBox, hash } = {},
    width = '24',
    height = '24',
    classes = 'fill-current',
    ...otherProps
}) => {
    const { store } = useContext(AppContext);

    return (
        <svg {...otherProps} class={classes} width={width} height={height} {...(viewBox ? { viewBox } : {})}>
            <use href={`${store.filenames?.sprite}${hash}`} />
        </svg>
    );
};
