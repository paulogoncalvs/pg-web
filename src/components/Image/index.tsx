import { h, FunctionalComponent } from 'preact';
import { useState, useRef } from 'preact/hooks';
import classNames from 'classnames';
import { useIntersectionObserver } from '@/hooks/useIntersectionObserver';

export const Image: FunctionalComponent<{
    isLazy?: boolean;
    src: string;
    srcset?: string;
    fallbackSrc?: string;
    width?: string;
    height?: string;
    classes?: string;
    alt?: string;
    style?: string;
    otherProps?: unknown;
}> = ({
    isLazy,
    src,
    srcset,
    fallbackSrc = 'data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7',
    classes,
    ...otherProps
}) => {
    const [isLoaded, setIsLoaded] = useState(false);
    const [isLoading, setIsLoading] = useState(false);

    const imgRef = useRef<HTMLImageElement>(null);
    const [entry, observer] = useIntersectionObserver(imgRef, { threshold: 0.25, root: null });
    const isVisible = !!entry?.isIntersecting;

    if (!isLoaded && !isLoading && isVisible) {
        fetch(src)
            .then(() => {
                setIsLoading(false);
                setIsLoaded(true);
                entry && observer?.unobserve(entry.target);
            })
            .catch((err) => {
                setIsLoading(false);
                setIsLoaded(false);
                console.error(err); // eslint-disable-line no-console
            });
    }

    return (
        <img
            ref={imgRef}
            {...otherProps}
            class={classNames(
                {
                    lazy: isLazy,
                    '_loaded animate-fade-in opacity-0': isLoaded,
                    'animate-pulse bg-gray-200 dark:bg-gray-900': !isLoaded,
                },
                classes,
            )}
            src={isLazy ? (isLoaded ? src : fallbackSrc) : src}
            data-src={isLazy ? (isLoaded ? src : undefined) : undefined}
            srcset={!isLazy ? srcset : undefined}
        />
    );
};
