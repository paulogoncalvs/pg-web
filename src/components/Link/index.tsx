import { h, FunctionalComponent } from 'preact';

interface LinkComponentProps {
    href?: string;
    classes?: string;
    newWindow?: boolean;
    otherProps?: unknown;
}

export const Link: FunctionalComponent<LinkComponentProps> = ({ href, classes, newWindow, children }) => (
    <a href={href} class={classes} {...(newWindow ? { target: '_blank', rel: 'noopener noreferrer' } : {})}>
        {children}
    </a>
);
