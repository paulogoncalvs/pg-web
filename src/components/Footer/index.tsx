import { h, FunctionalComponent } from 'preact';
import { Icon } from '@/components/Icon';
import { Link } from '@/components/Link';
import { t } from '@/modules/i18n';
import preactLogo from '@/assets/icons/logos/preact.svg';
import webpackLogo from '@/assets/icons/logos/webpack.svg';
import tailwindLogo from '@/assets/icons/logos/tailwind.svg';
import babelLogo from '@/assets/icons/logos/babel.svg';
import esLintLogo from '@/assets/icons/logos/eslint.svg';
import postCSSLogo from '@/assets/icons/logos/postcss.svg';
import prettierLogo from '@/assets/icons/logos/prettier.svg';
import sassLogo from '@/assets/icons/logos/sass.svg';
import styleLintLogo from '@/assets/icons/logos/stylelint.svg';
import typeScriptLogo from '@/assets/icons/logos/typescript.svg';
import yarnLogo from '@/assets/icons/logos/yarn.svg';
import workboxLogo from '@/assets/icons/logos/workbox.svg';
import jestLogo from '@/assets/icons/logos/jest.svg';

const initialYear = 2021;
const currentYear = new Date().getFullYear();

export const Footer: FunctionalComponent = () => (
    <footer>
        <div class="bg-gray-100 dark:bg-gray-900 text-center">
            <div class="container mx-auto px-6 pt-14 pb-12 flex flex-col items-center">
                <div class="pb-6 text-gray-600 dark:text-gray-400 text-sm sm:text-base">
                    {t('footer_description_1')}
                </div>
                <div class="flex flex-wrap justify-center align-middle">
                    <Link href="https://www.typescriptlang.org/" classes="ic-link _ftr" newWindow>
                        <Icon src={typeScriptLogo} aria-label="TypeScript" width="32" height="32" />
                    </Link>
                    <Link href="https://preactjs.com/" classes="ic-link _ftr" newWindow>
                        <Icon src={preactLogo} aria-label="Preact" width="32" height="32" />
                    </Link>
                    <Link href="https://webpack.js.org/" classes="ic-link _ftr" newWindow>
                        <Icon src={webpackLogo} aria-label="Webpack" width="32" height="32" />
                    </Link>
                    <Link href="https://yarnpkg.com/" classes="ic-link _ftr" newWindow>
                        <Icon src={yarnLogo} aria-label="Yarn" width="32" height="32" />
                    </Link>
                    <Link href="https://eslint.org/" classes="ic-link _ftr" newWindow>
                        <Icon src={esLintLogo} aria-label="ESLint" width="32" height="32" />
                    </Link>
                    <Link href="https://stylelint.io/" classes="ic-link _ftr" newWindow>
                        <Icon src={styleLintLogo} aria-label="StyleLint" width="32" height="32" />
                    </Link>
                    <Link href="https://developers.google.com/web/tools/workbox" classes="ic-link _ftr" newWindow>
                        <Icon src={workboxLogo} aria-label="Workbox" width="32" height="32" />
                    </Link>
                    <Link href="https://tailwindcss.com/" classes="ic-link _ftr" newWindow>
                        <Icon src={tailwindLogo} aria-label="Tailwind CSS" width="32" height="32" />
                    </Link>
                    <Link href="https://prettier.io/" classes="ic-link _ftr" newWindow>
                        <Icon src={prettierLogo} aria-label="Prettier" width="32" height="32" />
                    </Link>
                    <Link href="https://jestjs.io/" classes="ic-link _ftr" newWindow>
                        <Icon src={jestLogo} aria-label="Jest" width="32" height="32" />
                    </Link>
                    <Link href="https://postcss.org/" classes="ic-link _ftr" newWindow>
                        <Icon src={postCSSLogo} aria-label="PostCSS" width="32" height="32" />
                    </Link>
                    <Link href="https://sass-lang.com/" classes="ic-link _ftr" newWindow>
                        <Icon src={sassLogo} aria-label="Sass" width="32" height="32" />
                    </Link>
                    <Link href="https://babeljs.io/" classes="ic-link _ftr" newWindow>
                        <Icon src={babelLogo} aria-label="Babel" width="60" height="32" />
                    </Link>
                </div>
            </div>
        </div>
        <div class="container mx-auto py-8 flex flex-col items-center text-center text-gray-500 dark:text-gray-400 text-xs">
            <p>{t('footer_description_2')}</p>
            <p class="pt-2">
                paulogoncalves.dev &copy; {initialYear} {currentYear > initialYear ? ` - ${currentYear}` : ''} ðŸ™‚
            </p>
        </div>
    </footer>
);
