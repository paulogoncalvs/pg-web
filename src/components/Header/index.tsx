import { h, FunctionalComponent } from 'preact';
import { ToggleTheme } from '@/components/ToggleTheme';
import { Icon } from '@/components/Icon';
import { Link } from '@/components/Link';
import LanguageSelector from '@/components/LanguageSelector';
import linkedInIcon from '@/assets/icons/logos/linkedin.svg';
import githubIcon from '@/assets/icons/logos/github.svg';
import twitterIcon from '@/assets/icons/logos/twitter.svg';

export const Header: FunctionalComponent = () => (
    <header class="container mx-auto p-4 py-10">
        <div class="flex items-center justify-between">
            <div class="flex items-center">
                <LanguageSelector classes="ml-2 mr-2" />
                <ToggleTheme classes="p-2" />
            </div>
            <div class="flex justify-end">
                <Link href="https://www.twitter.com/paulogoncalvs" classes="ic-link _hdr" newWindow>
                    <Icon src={twitterIcon} aria-label="Twitter" />
                </Link>
                <Link href="https://www.github.com/paulogoncalvs" classes="ic-link _hdr" newWindow>
                    <Icon src={githubIcon} aria-label="GitHub" />
                </Link>
                <Link href="https://pt.linkedin.com/in/paulogoncalvs" classes="ic-link _hdr" newWindow>
                    <Icon src={linkedInIcon} aria-label="LinkedIn" />
                </Link>
            </div>
        </div>
    </header>
);
