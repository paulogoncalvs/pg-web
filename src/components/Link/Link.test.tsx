import { h } from 'preact';
import { render } from 'preact-render-to-string';
import { Link } from '@/components/Link';

describe('Link', () => {
    it('should render Link component', () => {
        expect(render(<Link href="/test/" />)).toMatchSnapshot();
    });

    it('should render Link component that opens on a new window', () => {
        expect(render(<Link href="/test/" newWindow />)).toMatchSnapshot();
    });
});
