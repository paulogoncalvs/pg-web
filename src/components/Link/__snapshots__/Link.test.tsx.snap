// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link should render Link component 1`] = `"<a href=\\"/test/\\"></a>"`;

exports[`Link should render Link component that opens on a new window 1`] = `"<a href=\\"/test/\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\"></a>"`;
